README
These is the readme file for the Survey extension.

Extension page on mediawiki.org: https://www.mediawiki.org/wiki/Extension:Survey

== About ==

Survey is a MediaWiki extension that allows administrators to define surveys
that can then be shown to specified groups of users.

=== Feature overview ===

* Surveys get displayed to wiki users using a nice pop-up form.
* Surveys can be created and modified via an admin panel (Special:Surveys).
* Surveys can be shown to specific types of users (logged in, anonymous, editor, confirmed, ...)
* Surveys can be shown for a specified percentage of users, and only after a certain amount of page views.
* Survey questions can have different types of inputs, of which some allow restricting answers.
* Surveys can be tested by admins via Special:TakeSurvey without them being live on the wiki.
* Summaries of survey submissions can be found on Special:SurveyStats.
* The survey submission data (ie user answers) can be exported via an API.

== Contributing ==

If you have bug reports or requests, please add them to the Talk page:
https://www.mediawiki.org/w/index.php?title=Extension_talk:Survey
You can also send them to Jeroen De Dauw by email, jeroendedauw at googles mail service.

For more info, see https://www.mediawiki.org/wiki/Extension:Survey

== Credits to other projects ==

* Fancybox
This extension includes code from the FancyBox library. FancyBox is an open-source
jQuery library dual licensed under the MIT [0] and GPL [1] licenses.

* jquery.numeric
This extension includes a copy of the jquery.numeric extension by Sam Collett.
It's dual licensed under the MIT [0] and GPL [1] licenses.

* JSON.js
This extension icnludes a copy of JSON.js, which is in the public domain.

[0] http://www.opensource.org/licenses/mit-license.php
[1] http://www.opensource.org/licenses/gpl-license.php


RELEASE-NOTES
These are the release notes for the Survey extension.

Extension page on mediawiki.org: https://www.mediawiki.org/wiki/Extension:Survey
Latest version of the release notes: https://gerrit.wikimedia.org/r/gitweb?p=mediawiki/extensions/Survey.git;a=blob;f=RELEASE-NOTES


=== Version 0.2.1 ===
(dev)

* Dropped compatibility with MediaWiki 1.18.
* Use of modern message functions.
* Use of local context instead of globals.

=== Version 0.2 ===
2012-9-24

* Dropped compatibility with MediaWiki 1.17.
* Use of RequestContext methods to get rid of globals.
* Cleanup of database table index code.
* Added missing database table indexes.
* Use of jquery.json for greater compatibility.

=== Version 0.1.2 ===
2011-11-14

* Fixed issue with database index names causing update.php to fail when running after installation of Survey.
* Fixed several IE JavaScript issues.
* Added missing action- messages.

=== Version 0.1.1 ===
2011-09-18

* Restricted access to survey statistics via the API to survey admins.

=== Version 0.1 ===
2011-09-13

Initial release with these features:

* Surveys get displayed to wiki users using a nice pop-up form.
* Surveys can be created and modified via an admin panel (Special:Surveys).
* Surveys can be shown to specific types of users (logged in, anonymous, editor, confirmed, ...)
* Surveys can be shown for a specified percentage of users, and only after a certain amount of page views.
* Survey questions can have different types of inputs, of which some allow restricting answers.
* Surveys can be tested by admins via Special:TakeSurvey without them being live on the wiki.
* Summaries of survey submissions can be found on Special:SurveyStats.
* The survey submission data (ie user answers) can be exported via an API.


Survey.php
<?php
/**
 * Initialization file for the Survey extension.
 *
 * Documentation:	 		https://www.mediawiki.org/wiki/Extension:Survey
 * Support					https://www.mediawiki.org/wiki/Extension_talk:Survey
 * Source code:			    https://gerrit.wikimedia.org/r/gitweb?p=mediawiki/extensions/Survey.git
 *
 * @file Survey.php
 * @ingroup Survey
 *
 * @licence GNU GPL v2+
 * @author Jeroen De Dauw < jeroendedauw@gmail.com >
 */

/**
 * This documentation group collects source code files belonging to Survey.
 *
 * @defgroup Survey Survey
 */

if ( !defined( 'MEDIAWIKI' ) ) {
	die( 'Not an entry point.' );
}

if ( version_compare( $wgVersion, '1.19c', '<' ) ) {
	die( '<b>Error:</b> This version of Survey requires MediaWiki 1.19 or above. For MediaWiki 1.18.x, you can use Survey 0.2.' );
}

define( 'Survey_VERSION', '0.3.0 alpha' );

$wgExtensionCredits['other'][] = array(
	'path' => __FILE__,
	'name' => 'Survey',
	'version' => Survey_VERSION,
	'author' => array(
		'[https://www.mediawiki.org/wiki/User:Jeroen_De_Dauw Jeroen De Dauw] for [http://www.wikiworks.com WikiWorks]',
	),
	'url' => 'https://www.mediawiki.org/wiki/Extension:Survey',
	'descriptionmsg' => 'survey-desc'
);

$wgMessagesDirs['Survey'] = __DIR__ . '/i18n';
$wgExtensionMessagesFiles['SurveyAlias']		= dirname( __FILE__ ) . '/Survey.alias.php';

$wgAutoloadClasses['SurveyHooks'] 				= dirname( __FILE__ ) . '/Survey.hooks.php';
$wgAutoloadClasses['SurveySettings'] 			= dirname( __FILE__ ) . '/Survey.settings.php';

$wgAutoloadClasses['ApiAddSurvey'] 				= dirname( __FILE__ ) . '/api/ApiAddSurvey.php';
$wgAutoloadClasses['ApiDeleteSurvey'] 			= dirname( __FILE__ ) . '/api/ApiDeleteSurvey.php';
$wgAutoloadClasses['ApiEditSurvey'] 			= dirname( __FILE__ ) . '/api/ApiEditSurvey.php';
$wgAutoloadClasses['ApiQuerySurveyAnswers'] 	= dirname( __FILE__ ) . '/api/ApiQuerySurveyAnswers.php';
$wgAutoloadClasses['ApiQuerySurveys'] 			= dirname( __FILE__ ) . '/api/ApiQuerySurveys.php';
$wgAutoloadClasses['ApiQuerySurveySubmissions']	= dirname( __FILE__ ) . '/api/ApiQuerySurveySubmissions.php';
$wgAutoloadClasses['ApiSubmitSurvey'] 			= dirname( __FILE__ ) . '/api/ApiSubmitSurvey.php';

$wgAutoloadClasses['Survey'] 					= dirname( __FILE__ ) . '/includes/Survey.class.php';
$wgAutoloadClasses['SurveyAnswer']	 			= dirname( __FILE__ ) . '/includes/SurveyAnswer.php';
$wgAutoloadClasses['SurveyCompat']	 			= dirname( __FILE__ ) . '/includes/SurveyCompat.php';
$wgAutoloadClasses['SurveyDBClass']	 			= dirname( __FILE__ ) . '/includes/SurveyDBClass.php';
$wgAutoloadClasses['SurveyQuestion'] 			= dirname( __FILE__ ) . '/includes/SurveyQuestion.php';
$wgAutoloadClasses['SurveySubmission'] 			= dirname( __FILE__ ) . '/includes/SurveySubmission.php';
$wgAutoloadClasses['SurveyTag'] 				= dirname( __FILE__ ) . '/includes/SurveyTag.php';

$wgAutoloadClasses['SpecialSurvey'] 			= dirname( __FILE__ ) . '/specials/SpecialSurvey.php';
$wgAutoloadClasses['SurveyQuestionField']		= dirname( __FILE__ ) . '/specials/SpecialSurvey.php';
$wgAutoloadClasses['SurveyNameField']			= dirname( __FILE__ ) . '/specials/SpecialSurvey.php';
$wgAutoloadClasses['SpecialSurveyPage'] 		= dirname( __FILE__ ) . '/specials/SpecialSurveyPage.php';
$wgAutoloadClasses['SpecialSurveys'] 			= dirname( __FILE__ ) . '/specials/SpecialSurveys.php';
$wgAutoloadClasses['SpecialSurveyStats'] 		= dirname( __FILE__ ) . '/specials/SpecialSurveyStats.php';
$wgAutoloadClasses['SpecialTakeSurvey'] 		= dirname( __FILE__ ) . '/specials/SpecialTakeSurvey.php';

$wgSpecialPages['EditSurvey'] 					= 'SpecialSurvey';
$wgSpecialPages['Surveys'] 						= 'SpecialSurveys';
$wgSpecialPages['SurveyStats'] 					= 'SpecialSurveyStats';
$wgSpecialPages['TakeSurvey'] 					= 'SpecialTakeSurvey';

$wgAPIModules['addsurvey'] 						= 'ApiAddSurvey';
$wgAPIModules['deletesurvey'] 					= 'ApiDeleteSurvey';
$wgAPIModules['editsurvey'] 					= 'ApiEditSurvey';
$wgAPIModules['submitsurvey'] 					= 'ApiSubmitSurvey';
$wgAPIListModules['surveys'] 					= 'ApiQuerySurveys';
$wgAPIListModules['surveyanswers'] 				= 'ApiQuerySurveyAnswers';
$wgAPIListModules['surveysubmissions'] 			= 'ApiQuerySurveySubmissions';

$wgHooks['LoadExtensionSchemaUpdates'][] 		= 'SurveyHooks::onSchemaUpdate';
$wgHooks['UnitTestsList'][] 					= 'SurveyHooks::registerUnitTests';
$wgHooks['ParserFirstCallInit'][] 				= 'SurveyHooks::onParserFirstCallInit';
$wgHooks['ArticleViewHeader'][] 				= 'SurveyHooks::onArticleViewHeader';
$wgHooks['AdminLinks'][] 						= 'SurveyHooks::addToAdminLinks';

$wgAvailableRights[] = 'surveyadmin';
$wgAvailableRights[] = 'surveysubmit';

# Users that can manage the surveys.
$wgGroupPermissions['*'            ]['surveyadmin'] = false;
$wgGroupPermissions['user'         ]['surveyadmin'] = false;
$wgGroupPermissions['autoconfirmed']['surveyadmin'] = false;
$wgGroupPermissions['bot'          ]['surveyadmin'] = false;
$wgGroupPermissions['sysop'        ]['surveyadmin'] = true;

# Users that can submit surveys.
$wgGroupPermissions['*'            ]['surveysubmit'] = true;
$wgGroupPermissions['user'         ]['surveysubmit'] = true;
$wgGroupPermissions['autoconfirmed']['surveysubmit'] = true;
$wgGroupPermissions['bot'          ]['surveysubmit'] = false;
$wgGroupPermissions['sysop'        ]['surveysubmit'] = true;

$moduleTemplate = array(
	'localBasePath' => dirname( __FILE__ ) . '/resources',
	'remoteExtPath' => 'Survey/resources'
);

$wgResourceModules['ext.survey'] = $moduleTemplate + array(
	'scripts' => array(
		'ext.survey.js'
	),
);

$wgResourceModules['ext.survey.special.surveys'] = $moduleTemplate + array(
	'scripts' => array(
		'ext.survey.special.surveys.js'
	),
	'dependencies' => array( 'ext.survey' ),
	'messages' => array(
		'surveys-special-confirm-delete',
		'surveys-special-delete-failed',
	)
);

$wgResourceModules['ext.survey.special.survey'] = $moduleTemplate + array(
	'scripts' => array(
		'ext.survey.answerSelector.js',
		'ext.survey.special.survey.js'
	),
	'styles' => array(
		'ext.survey.special.survey.css'
	),
	'dependencies' => array( 'ext.survey', 'jquery.ui.button' ),
	'messages' => array(
		'survey-question-type-text',
		'survey-question-type-number',
		'survey-question-type-select',
		'survey-question-type-radio',
		'survey-question-type-textarea',
		'survey-question-type-check',
		'survey-question-label-nr',
		'survey-special-label-required',
		'survey-special-label-type',
		'survey-special-label-text',
		'survey-special-label-addquestion',
		'survey-special-label-add-first',
		'survey-special-label-add-another',
		'survey-special-remove',
		'survey-special-remove-confirm',
		'survey-special-label-answers'
	)
);

$wgResourceModules['ext.survey.numeric'] = $moduleTemplate + array(
	'scripts' => array(
		'jquery.numeric.js'
	)
);

$wgResourceModules['ext.survey.jquery'] = $moduleTemplate + array(
	'scripts' => array(
		'fancybox/jquery.fancybox-1.3.4.js',
		'jquery.survey.js'
	),
	'styles' => array(
		'jquery.survey.css',
		'fancybox/jquery.fancybox-1.3.4.css'
	),
	'dependencies' => array(
		'ext.survey',
		'jquery.ui.button',
		'ext.survey.numeric',
		version_compare( $GLOBALS['wgVersion'], '1.24', '<' ) ? 'jquery.json' : 'json'
	),
	'messages' => array(
		'survey-jquery-submit',
		'survey-jquery-finish',
		'survey-jquery-load-failed',
	)
);

$wgResourceModules['ext.survey.tag'] = $moduleTemplate + array(
	'scripts' => array(
		'ext.survey.tag.js'
	),
	'dependencies' => array( 'ext.survey.jquery', 'jquery.cookie' ),
);

unset( $moduleTemplate );

$egSurveySettings = array();
